### PR TITLE
Bug/1849 attribute values and multiple accept types

### PR DESCRIPTION
**Pull request by kzangeli**
Fixes issues #1037, #1849, #1886, #2175, and #2208

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -30,3 +30,4 @@
 - Fix: Avoid rendering at the same time id and idPattern fields in NGSIv2 subscriptions
 - Fix: Error returned on GET/DELETE requests specifying Content-Type (Issue #2128)
 - Fix: Error responses with '405 Method Not Allowed' now return payload to describe the error, for v2 requests only (Issue #2075, #2078 and #2083)
+- Fix: Implemented more sofisticated HTTP Accept Header handling (Issue #1849)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: The five issue numbers solved by this PR should be mentioned.

In addition, not sure about "more sofisticated" is the best alternative... It is not about sofistication, but about correctness or accuracy.

File: src/lib/rest/HttpHeaders.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  acceptHeaderV.clear();
+}
+
+
+
+/* ****************************************************************************
+*
+* HttpHeaders::accepted - 
+*/
+bool HttpHeaders::accepted(const std::string& mime)
```
Comment (fgalan) [inline src/lib/rest/HttpHeaders.cpp]: What about `*/json` and `*/plain`?

Not sure how to deal with them... the "* thing" should be checked in the MIME spec.

File: src/lib/rest/HttpHeaders.h [hunk 3]
```diff
@@ -26,6 +26,37 @@
 * Author: Ken Zangelin
 */
 #include <string>
+#include <vector>
+
+#include "common/MimeType.h"
+
+
+
+// -----------------------------------------------------------------------------
```
Comment (fgalan) [inline src/lib/rest/HttpHeaders.h]: We typically use `/* ... */` for preambles. Not really important, but it has grabed my attention...

File: src/lib/rest/rest.cpp [hunk 4]
```diff
@@ -262,18 +262,236 @@ static int uriArgumentGet(void* cbDataP, MHD_ValueKind kind, const char* ckey, c
...
+    ciP->httpStatusCode    = SccBadRequest;
+    ciP->outMimeType       = mimeTypeSelect(ciP, "Error");
+    delete acceptHeaderP;
+    return false;
+  }
+
+
+  //
+  // And now the accept-extensions  (which is ignored for now)
+  // FIXME P4: Implement treatment of accept-extensions
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: What about accept-params? Are we doing something with them? In negative case, it deserves a similar FIXME line.

File: src/lib/rest/HttpHeaders.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/rest/HttpHeaders.cpp]: */json and */plain are not valid.
only application/* and text/*

NTC

File: test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test [hunk 5]
```diff
@@ -181,12 +181,12 @@ Date: REGEX(.*)
...
 ==============================================
 HTTP/1.1 200 OK
-Content-Length: 11
+Content-Length: 13
 Content-Type: text/plain
 Fiware-Correlator: REGEX([0-9a-f\-]{36})
 Date: REGEX(.*)
 
-V1 A1 E1/T1
+"V1 A1 E1/T1"
```
Comment (fgalan) [inline test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test]: This is another fix different than the one already described in the CNR about Accept header processing. I sugges to describe it in an additional entry in CNR (in this case, without associated issue number, as far as I remember).

File: src/lib/rest/HttpHeaders.h (hunk 3 above)
Comment (kzangeli) [inline src/lib/rest/HttpHeaders.h]: :-)
My mistake

File: src/lib/rest/rest.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: q=  is accept-params. There should be a comment somewhere about the anatomy of an 'accept item'

File: test/functionalTest/cases/1849_attribute_values_and_multiple_accept_types/attribute_values_and_multiple_accept_types_for_number_attribute.test [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+# 08. GET I with Accept header text/*;q=0.3,application/json;q=0.8 - see plain text (q is overridden)
+# 09. GET I with Accept header application/json - see error
+#
+
+echo "01. Create E1 with an attribute of each type (S: string, I: number, B: bool, N: null, O: json object, A: json array)"
+echo "===================================================================================================================="
+payload='{
+  "id": "E1",
+  "S": "S",
+  "I": 3.14,
```
Comment (fgalan) [inline test/functionalTest/cases/1849_attribute_values_and_multiple_accept_types/attribute_values_and_multiple_accept_types_for_number_attribute.test]: I guess that "I" was intended for Integer, but at the end the number is decimal... Maybe D should be more clear.

Actually not very important (you would need to touch a lot of .test files to implemnt that change, so probably it doesn't worth the effort).

File: test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test]: Ah true. This is a 'bug' that I fixed on the way ...

File: src/lib/rest/HttpHeaders.h (hunk 3 above)
Comment (kzangeli) [inline src/lib/rest/HttpHeaders.h]: Wait, this was no mistake.

The text in the comment contains  `*/*` and thus would end the C-comment right there, resulting in a compilation error.
That's why I used C++ comments here.

NTC

File: src/lib/rest/rest.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: The comment header of HttpAcceptHeader in HttpHeaders.h - that's where the whole thing is explained.

NTC

File: test/functionalTest/cases/1849_attribute_values_and_multiple_accept_types/attribute_values_and_multiple_accept_types_for_number_attribute.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/1849_attribute_values_and_multiple_accept_types/attribute_values_and_multiple_accept_types_for_number_attribute.test]: I wanted to use N, but N I needed for Null.
I took I instead. Should have picked D instead, perhaps.
As you say, not really worth the effort.

NTC

File: test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0951_get_v2_param_type/get_v2_param_type.test]: Fixed in 03d1ed1

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in 03d1ed1

**Pull request merged**